### PR TITLE
Make it possible for self.forward to return a ScriptMethod

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19603 Serialize first-class version of functions
* #19371 Trace directly into first-class module form.
* #19334 @torch.jit.script(fn) now is a torch.jit.Function
* #19333 pybind CompilationUnit and Function directly
* **#19217 Make it possible for self.forward to return a ScriptMethod**

Summary:
Previously we had to have a bunch of special case handling
for the 'forward' method because it was already defined in nn.Module,
making it not go through __getattr__. This commit adds a work around.

It also makes things like foo.forward.code or foo.forward.graph work.

To avoid cyclic references in this solution, this also splits
ScriptModule (the python class) and torch._C.ScriptModule the pybind11 wrapper
on the C++ class. I think this change actually makes the code easier
to understand, because it avoids weird case where both _C.ScriptModule
and ScriptModule both need meta-classes to work.

By caching methods, we avoid two trips through C++ in the steady state.

Test Plan: python test/test_jit.py

Differential Revision: [D14922120](https://our.internmc.facebook.com/intern/diff/D14922120)

### CODE REVIEW

File: torch/csrc/jit/script/init.cpp
Comment (ZolotukhinM) [inline torch/csrc/jit/script/init.cpp]: Add `AT_ASSERT(mod)` ?

File: test/test_jit.py
Comment (ZolotukhinM) [inline test/test_jit.py]: Why don't we need to run super's init here?

File: test/test_jit.py
Comment (ZolotukhinM) [inline test/test_jit.py]: Nit: this seems to be an unrelated change.

**Review by ZolotukhinM — APPROVED** [review]
The changes look reasonable, but I need to admit I didn't understand all the deep details. Please feel free to commit if you are certain or wait for another review from someone else.

File: test/test_jit.py
Comment (zdevito) [reply test/test_jit.py]: If there is no `__init__` then you just inherit the super's `__init__`. If you define your _own_ function, then you must call super. Not the best-designed thing in python...

File: test/test_jit.py
Comment (zdevito) [reply test/test_jit.py]: It is related. Separating the C/Python type made this test fail because it forgot to call super. I fixed the tests where this was the case and our library is much more verbose about saying that you forgot to call __init__ now.

**Pull request closed without merging**
